### ISSUE TITLE
mccabe dependency update causing installation failures

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#557]
In GitLab by @ashcrow on Jan 23, 2017, 10:49

## Installation
```python setup.py flake8```

See: https://github.com/openshift/openshift-ansible/pull/3141#issuecomment-274559282

## Problem
```flake8``` may fail to install now that  [mccabe 0.6.0 ](https://pypi.python.org/pypi/mccabe/) has been released.

## Output
```
Traceback (most recent call last):
  File "setup.py", line 77, in <module>
    'oo-install=ooinstall.cli_installer:cli',
  File "/usr/lib64/python2.7/distutils/core.py", line 137, in setup
    ok = dist.parse_command_line()
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/setuptools/dist.py", line 358, in parse_command_line
    result = _Distribution.parse_command_line(self)
  File "/usr/lib64/python2.7/distutils/dist.py", line 467, in parse_command_line
    args = self._parse_command_opts(parser, args)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/setuptools/dist.py", line 669, in _parse_command_opts
    nargs = _Distribution._parse_command_opts(self, parser, args)
  File "/usr/lib64/python2.7/distutils/dist.py", line 523, in _parse_command_opts
    cmd_class = self.get_command_class(command)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/setuptools/dist.py", line 489, in get_command_class
    ep.require(installer=self.fetch_build_egg)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2306, in require
    items = working_set.resolve(reqs, env, installer)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/pkg_resources/__init__.py", line 853, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (mccabe 0.6.0 (/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages), Requirement.parse('mccabe<0.6.0,>=0.5.0'))
```

## Fix

Looks like the [fix is already in HEAD](https://gitlab.com/pycqa/flake8/blob/master/setup.py#L22). However, those using pypi are still pulling the current release which requires versions under ``0.6.0``.

See https://gitlab.com/ashcrow/flake8/commit/83333217512d7cff3a9f71bbcee23bff38dd6fb0

**Comment by asottile** [comment]
In GitLab by @ashcrow on Jan 24, 2017, 06:51

FWIW here is the changelog for 0.6.0

```
0.6.0 - 2017-01-23
Add support for Python 3.6
Fix handling for missing statement types
```

It *sounds* compatible with previous versions based off that.

**Comment by asottile** [comment]
In GitLab by @ashcrow on Jan 24, 2017, 07:22

I updated the requirements to allow ```mccabe``` 0.6.0 locally and tested.

## Python 2

```mccabe``` 0.6.0 seems to work fine with ```flake8``` 3.2.1.

```terminal
$ virtualenv t
New python executable in /tmp/t/bin/python2
Also creating executable in /tmp/t/bin/python
Installing setuptools, pip, wheel...done.
$ . t/bin/activate
$ pip install -U mccabe
Collecting mccabe
  Using cached mccabe-0.6.0-py2.py3-none-any.whl
Installing collected packages: mccabe
Successfully installed mccabe-0.6.0
$ wget https://pypi.python.org/packages/51/70/af1b3130cd051f2f79854b14079b3bcbad84b8bac31a7dffc63ef57f8a7a/flake8-3.2.1.tar.gz#md5=06b388714877a1f758434379c3a94660
<snip>
$ tar xvf flake8-3.2.1.tar.gz 
<snip>
$ cd flake8-3.2.1
$ python setup.py install
<snip>
$ pip freeze | grep flake8
flake8==3.2.1
$ pip freeze | grep mccabe
mccabe==0.6.0
$ flake8 src/
$ flake8 --max-complexity 9 src/
src/flake8/utils.py:209:1: C901 'filenames_from' is too complex (10)
$ flake8 --max-complexity 5 src/ | wc -l
12
(t) [steve@bitfall flake8-3.2.1]$ echo $?
0
$
```

## Python 3

```mccabe``` 0.6.0 seems to work fine with ```flake8``` 3.2.1.

**NOTE**: I used the same ```wget``` download from the Python 2 test.

```terminal
$ virtualenv-3 t
Using base prefix '/usr'
New python executable in /tmp/t/bin/python3
Also creating executable in /tmp/t/bin/python
Installing setuptools, pip, wheel...done.
$ . t/bin/activate
$ python setup.py install
$ pip freeze | grep flake8
flake8==3.2.1
$ pip freeze | grep mccabe
mccabe==0.6.0
$ flake8 src/
$ flake8 --max-complexity 9 src/
src/flake8/utils.py:209:1: C901 'filenames_from' is too complex (10)
$  flake8 --max-complexity 5 src/ | wc -l
13
$ echo $?
0
```

**Comment by asottile** [comment]
In GitLab by @ashcrow on Jan 24, 2017, 07:36

Created https://gitlab.com/ashcrow/flake8/commit/83333217512d7cff3a9f71bbcee23bff38dd6fb0 which is based off the ``3.2.1`` tag. This is the change I used to test with.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 25, 2017, 05:36

closed

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 25, 2017, 05:36

Hi @ashcrow, thanks for the report!

So I looked at this with @jaraco yesterday and he helped me determine the cause of this for you.

First, let's start with some history. We explain [here](http://flake8.pycqa.org/en/latest/faq.html#why-does-flake8-use-ranges-for-its-dependencies) why we cap dependencies, so I won't re-explain that. In short, however, mccabe 0.6 includes changes to computing complexity, so, if Flake8 were to leave it uncapped some people would have suddenly different behaviour in their CI environment before and after 0.6.0 was released. This is why we only change the caps/ranges around minor versions of Flake8 (3.2, 3.3, etc.).

Second, many things depend on mccabe, including PyLint. Not all of those things will cap mccabe. If Flake8 were your only dependency using mccabe, then you wouldn't have this problem. Alas, you have at least one other dependency that does require it. Pip will install a version that matches the first constraint it finds. So if it comes across PyLint first, it'll install the newest mccabe available. If it comes across Flake8 first, then it would install something from the 0.5 series. In the former case, it doesn't downgrade once it sees Flake8 wants something earlier than 0.6.0.

Finally, the version conflict error you see comes from `pkg_resources`. On import, `pkg_resources` scans everything and enforces requirement constraints. This will eventually be turned off (enforcing requirement constraints) but it is presently on.

So, this is emphatically *not* a Flake8 bug. We cap our dependencies to provide the best, most consistent, user experience and we release when we *can*. (For example, pycodestyle still needs to generate a release that supports Python 3.6.) We will release 3.3.0 when those requirements are met and some of the items here that are targeted to 3.3.0 are finished.

Your "fix" in Flake8 is not going to be merged for a 3.2.x release. And your diagnosis was incomplete. Flake8 won't fail with mccabe 0.6.0 in the wild *unless* you force that version to be installed (either intentionally or not).

You can either reorder your requirements, or you can explicitly cap mccabe, or fix dependency resolution in pip, or help `pkg_resources` turn off implicit requirement checking. I doubt PyLint will bother to set caps on their dependencies because they probably don't have the patience to deal with the frequent issues being opened that they aren't using the latest version. I would suggest you reorder your requirements file, add a comment, and move on. Flake8 isn't going to change how it handles dependencies either because this truly produces the *least* number of surprises.
